sdk/lib/socks5-listener/src/lib.rs

```diff
@@ -107,13 +103,15 @@ pub fn get_client_state() -> ClientState {
     }
 }
 
-#[ffi_export]
-pub fn start_client(
+pub fn start_client<F, S>(
     storage_directory: Option<char_p::Ref<'_>>,
     service_provider: Option<char_p::Ref<'_>>,
-    on_start_callback: StartupCallback<'static>,
-    on_shutdown_callback: ShutdownCallback<'static>,
-) {
+    on_start_callback: F,
+    on_shutdown_callback: S,
+) where
+    F: FnMut(String) + Send + 'static,
+    S: FnMut() + Send + 'static,
```

Comment:
nitpicky: does it have to be an `FnMut` or could we get away with a normal `Fn` ( or maybe even something like we had initially, i.e. `extern "C" fn`)?

Reply:
`FnMut` is needed as for android we are using JNI to call java methods and the signature takes a `&mut self`
https://docs.rs/jni/0.21.1/jni/struct.JNIEnv.html#method.call_method

> or maybe even something like we had initially, i.e. extern "C" fn

Im not sure, can we share a snippet using `extern "C" fn` please?